mount/titus-mount-lustre.c

```diff
@@ -0,0 +1,359 @@
+	E_fsconfig(fsfd, FSCONFIG_CMD_CREATE, NULL, NULL, 0);
+}
+
+static int setup_fsfd_in_namespaces(int sk, int nsfd)
+{
+	int fsfd, ret;
+	int usernsfd = openat(nsfd, "user", O_RDONLY);
+	int mnt_fd = openat(nsfd, "mnt", O_RDONLY);
+	int net_fd = openat(nsfd, "net", O_RDONLY);
+	assert(usernsfd != -1);
+	assert(mnt_fd != -1);
+	assert(net_fd != -1);
+	ret = setns(usernsfd, CLONE_NEWUSER);
+	if (ret == -1) {
+		perror("setns user");
+		return 1;
+	}
+	ret = setns(mnt_fd, CLONE_NEWNS);
+	if (ret == -1) {
+		perror("setns mnt");
+		return 1;
+	}
+	ret = setns(net_fd, CLONE_NEWNET);
+	if (ret == -1) {
+		perror("setns net");
+		return 1;
+	}
+	fsfd = fsopen("lustre", 0);
+	if (fsfd == -1) {
+		perror("fsopen");
+		return 1;
+	}
+	assert(send_fd(sk, fsfd) == 0);
+	return 0;
+}
+
+static int fork_and_get_fsfd(long int nsfd)
+{
+	int sk_pair[2], ret, fsfd, status;
+	pid_t worker;
+
+	if (socketpair(PF_LOCAL, SOCK_SEQPACKET, 0, sk_pair) < 0) {
+		perror("socketpair");
+		exit(1);
+	}
+	worker = fork();
+	if (worker < 0) {
+		perror("fork");
+		exit(1);
+	}
+	if (worker == 0) {
+		close(sk_pair[0]);
+		ret = setup_fsfd_in_namespaces(sk_pair[1], nsfd);
+		close(sk_pair[1]);
+		exit(ret);
+	}
+	close(sk_pair[1]);
+	fsfd = recv_fd(sk_pair[0]);
+	assert(fsfd >= 0);
+	if (waitpid(worker, &status, 0) != worker) {
+		perror("waitpid");
+		exit(1);
+	}
+	if (!WIFEXITED(status) || WEXITSTATUS(status)) {
+		fprintf(stderr, "worker exited nonzero\n");
+		exit(1);
+	}
+	close(sk_pair[0]);
+	return fsfd;
+}
+
+static void switch_namespaces(int nsfd)
+{
+	int ret;
+	int mnt_fd = openat(nsfd, "mnt", O_RDONLY);
+	assert(mnt_fd != -1);
+	ret = setns(mnt_fd, CLONE_NEWNS);
+	if (ret == -1) {
+		perror("setns mount");
+		exit(1);
+	}
+	int net_fd = openat(nsfd, "net", O_RDONLY);
+	assert(net_fd != -1);
+	ret = setns(net_fd, CLONE_NEWNET);
+	if (ret == -1) {
+		perror("setns net");
+		exit(1);
+	}
+}
+
+static void mount_and_move(int fsfd, const char *target,
+			   unsigned long flags)
+{
+	int mfd = fsmount(fsfd, 0, flags);
+	if (mfd < 0) {
+		mount_error(fsfd, "fsmount");
+		E(close(fsfd));
+	}
+	E(move_mount(mfd, "", AT_FDCWD, target, MOVE_MOUNT_F_EMPTY_PATH));
+	E(close(mfd));
+}
+
+int main(int argc, char *argv[])
+{
+	int nsfd, fsfd, res = -1;
+	unsigned long flags_ul;
+	/*
+	 * We do this because parsing args is a bigger pain than passing
+	 * via environment variable, although passing via environment
+	 * variable has a "cost" in that they are limited in size
+	 */
+	const char *pidDir = getenv("TITUS_PID1_DIR");
+	const char *lustre_mount_hostname = getenv("MOUNT_LUSTRE_HOSTNAME");
+    const char *lustre_mount_mountname = getenv("MOUNT_LUSTRE_MOUNTNAME");
+	const char *target = getenv("MOUNT_TARGET");
+	const char *flags = getenv("MOUNT_FLAGS");
+	const char *options = getenv("MOUNT_OPTIONS");
+
+	int buf_size = sysconf(_SC_PAGESIZE);
+	char final_options[buf_size];
+	strcpy(final_options, options);
```

Comment:
`strncpy(final_options, options, buf_size)`?

Could also check length of `options` to see if it would exceed `_SC_PAGESIZE`. For context it looks like `_SC_PAGESIZE` would typically be 4096 so that'd be weird but not impossible

Reply:
We are controlling all the inputs here, we don't have to be too defensive